### PR TITLE
refactor mac for state machine implementation

### PR DESCRIPTION
**Pull request by lthiery**
This is the first stage in some refactoring that I am doing to try to unify the async and no_async (state_machine) device implementations. 

I've minimized impact on the async_device area of the code to try to keep things simple. However, I believe that this should allow us to move more non-runtime logic into the `mac` module and to share more of the protocol logic amongst the two runtimes.

I ran into this as I was working on a solution for #117 and I realized it would be a lot easier to start consolidating some of the duplicate code.

EDIT: to be clear, I think this refactor starts to make it easier to share the [logic here](https://github.com/ivajloip/rust-lorawan/blob/7f7ceef024991d0f068b59cff715897b021e8521/device/src/async_device/mod.rs#L311-L360).

### CODE REVIEW

File: device/src/lib.rs [hunk 1]
```diff
@@ -2,16 +2,16 @@
 #![cfg_attr(feature = "async", feature(async_fn_in_trait))]
 #![allow(incomplete_features)]
 
+extern crate alloc;
```
Comment (lulf) [inline device/src/lib.rs:5]: This should be behind a feature flag for the cases where alloc is not supported (most embassy examples are no alloc), but I also can't see any place in the code that requires it?

File: device/src/mac/mod.rs [hunk 2]
```diff
@@ -0,0 +1,178 @@
+use crate::{
+    radio::RadioBuffer, region, state_machines::Downlink, FcntDown, FcntUp, SendData, SessionKeys,
+};
+use generic_array::{typenum::U256, GenericArray};
```
Comment (lulf) [inline device/src/mac/mod.rs:4]: Just a note: it would be good to change this to use const generics instead of generic array at some point.

Comment (lthiery) [reply device/src/mac/mod.rs:4]: Good point. Left an issue to track it #126

File: device/src/lib.rs (hunk 1 above)
Comment (lthiery) [reply device/src/lib.rs]: Woah! Thanks for catching that. It's really not meant to be there. My editor does that sometimes and now I'm motivated to disable that feature.

**Comment by lthiery** [comment]
FYI, I'm also considering getting some tests setup around this one. It should be quite a bit easier than the async ones.

**Review by lulf — DISMISSED** [review]

File: device/src/mac/mod.rs [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+                        Some(Response::SessionExpired)
+                    } else {
+                        Some(Response::DownlinkReceived(fcnt))
+                    };
+                }
+            }
+        }
+        None
+    }
+    pub(crate) fn rx2_elapsed(&mut self) -> Response {
```
Comment (ivajloip) [inline device/src/mac/mod.rs:106]: nit: adding a new line above can improve readability.

**Review by ivajloip — APPROVED** [review]
I am getting tired - long day, but I tried to quickly scan this and it looks good, thanks! :)

**Review by lulf — APPROVED** [review]

**Pull request merged**